dbms/src/Storages/DeltaMerge/DeltaMergeStore_Ingest.cpp

```diff
@@ -210,6 +140,412 @@ void DeltaMergeStore::ingestFiles(
+                RUNTIME_CHECK(compare(segment_ingest_range.getEnd(), file_ingest_range.getStart()) > 0);
+                file_ingest_range.setStart(segment_ingest_range.end);
+            }
+            else
+            {
+                // this segment is abandoned, or may be split into multiples.
+                // retry with current range and file and find segment again.
+            }
+        }
+    }
+
+    LOG_FMT_DEBUG(
+        log,
+        "Table ingest using split - split ingest phase - finished, updated_segments_n={}",
+        updated_segments.size());
+
+    return std::vector<SegmentPtr>(
+        updated_segments.begin(),
+        updated_segments.end());
+}
+
+/**
+ * Ingest one DTFile into the target segment by using logical split.
+ */
+bool DeltaMergeStore::ingestDTFileIntoSegmentUsingSplit(
+    DMContext & dm_context,
+    const SegmentPtr & segment,
+    const RowKeyRange & ingest_range,
+    const DMFilePtr & file)
+{
+    const auto & segment_range = segment->getRowKeyRange();
+
+    // The ingest_range must fall in segment's range.
+    RUNTIME_CHECK(
+        !ingest_range.none(),
+        ingest_range.toDebugString());
+    RUNTIME_CHECK(
+        compare(segment_range.getStart(), ingest_range.getStart()) <= 0,
+        segment_range.toDebugString(),
+        ingest_range.toDebugString());
+    RUNTIME_CHECK(
+        compare(segment_range.getEnd(), ingest_range.getEnd()) >= 0,
+        segment_range.toDebugString(),
+        ingest_range.toDebugString());
+
+    const bool is_start_matching = (compare(segment_range.getStart(), ingest_range.getStart()) == 0);
+    const bool is_end_matching = (compare(segment_range.getEnd(), ingest_range.getEnd()) == 0);
+
+    if (is_start_matching && is_end_matching)
+    {
+        /*
+         * The segment and the ingest range is perfectly matched. We can
+         * simply replace all of the data from this segment.
+         *
+         * Example:
+         *    │----------- Segment ----------│
+         *    │-------- Ingest Range --------│
+         */
+        const auto new_segment_or_null = segmentDangerouslyReplaceData(dm_context, segment, file);
+        const bool succeeded = !!new_segment_or_null;
+        return succeeded;
+    }
+    else if (is_start_matching)
+    {
+        /*
+         * Example:
+         *    │--------------- Segment ---------------│
+         *    │-------- Ingest Range --------│
+         *
+         * We will logical split the segment to form a perfect matching segment:
+         *    │--------------- Segment ------│--------│
+         *    │-------- Ingest Range --------│
+         */
+        const auto [left, right] = segmentSplit(dm_context, segment, SegmentSplitReason::IngestBySplit, ingest_range.end, SegmentSplitMode::Logical);
+        // We don't care whether it is succeeded or not, because the caller always need to retry
+        // from the beginning in this case.
+        UNUSED(left);
+        UNUSED(right);
```

Comment:
```suggestion
        UNUSED(left, right);
```
`UNUSED` can accept multiple prarms